[RDY] Incsub 3

KillTheMule opened the pull request:
Incsubsitution feature

Originally implemented by
- Clement0
- DesbyP
- aym7
- Adrey06
- Robinhola

in #4811. Major reworkings and bug fixes by
- bfredl

Most tests suggested by ZyX-l, suggestions for improvements by oni-link and justinmk.

KillTheMule committed:
Incsubsitution feature

Originally implemented by

* Clement0
* DesbyP
* aym7
* Adrey06
* Robinhola

in #4811. Major reworkings and bug
fixes by

* bfredl

Most tests suggested by ZyX-l, suggestions for improvements by oni-link.

KillTheMule committed:
Tests for incsubstitution feature

KillTheMule commented:
Continuation of https://github.com/neovim/neovim/pull/4915, which became too large for my browser to properly handle, hope nobody minds.

Todo:

- Improve cursor placement during incsub (column!)
- Test cursor placement during incsub. Most will be done already when the screen tests are adjusted.
- Fix scrolling of the incsub window (note to self: `window_set_cursor`)
- Post an issue at vim for [this test](https://github.com/neovim/neovim/pull/5226/commits/595b2429b790b97785dc57cfebec6c9ee67c0be3#diff-4dbe255ebfc94f6664c66ffeab750809R842) which exhibits a potentially buggy behavior of standard subsitution (note to self: first test in vim itself, maybe it's an nvim issue).

bfredl commented on src/nvim/ex_cmds.c:
```diff
@@ -5864,3 +5928,204 @@ void set_context_in_sign_cmd(expand_T *xp, char_u *arg)
+
+  if (split) {
+    // Don't execute autocommands while creating the window.
+    block_autocmds();
```
I think these are redundant now we block autocommands in do_inc_sub.

KillTheMule replied:
Hmm there are 2 unblocks for this block below, wonder why this did not wreak havoc... thanks!

bfredl commented on src/nvim/ex_cmds.h:
```diff
@@ -26,6 +28,37 @@ typedef struct {
+
+// List of matches in a line
+#define _noop(x)
+KLIST_INIT(colnr_T, colnr_T, _noop)
```
I think we should use kvec here (and below) which is more efficient for an append-only sequence.

KillTheMule replied:
I checked that out a bit, it would mean hand-writing the for-loops, which is certainly doable. Also, I'd have to deallocate manually in the case of `MatchedLine`, since there's no way to initialize with a custom deallocator as with klist.

Should I still do it? I think I can manage, but I'm uncertain if deallocating manually is a good tradeoff.

KillTheMule commented:
I cleared up the autocmds stuff, there was also an explicit call for CMDWIN_LEAVE in there I removed (or is there any special reason for that one?).

I also fixed the cursor placement, though there is a catch: If you don't have `gdefault` set, then the incsub might find a match on the cursor's line _before_ and _after_ the cursor. In that case, the 2nd match doesn't count because we don't have the `g` flag while typing, so the cursor can't jump there and takes another one. It gets highlighted though, so that might be confusing. I'd say that's a problem of highlighting, but not cursor movement, and I'd also argue that this is probably the way it should be (but I'd also argue that `gdefault` is what should be the default..). Thoughts?

Still to be done:

- bfredl wants to check out if he can avoid the buffer numbers going up during incsub
- Post an issue at vim for [this test](https://github.com/neovim/neovim/pull/5226/commits/595b2429b790b97785dc57cfebec6c9ee67c0be3#diff-4dbe255ebfc94f6664c66ffeab750809R842) which exhibits a potentially buggy behavior of standard subsitution (note to self: first test in vim itself, maybe it's an nvim issue).
- Check out using `kvec` instead of `klist`

KillTheMule commented:
I also found a little bug in the highlighting of the replacement pattern in the incsub window, which I fixed and added a test for.

KillTheMule commented:
Here's a problem I'm uncertain how to deal with:

[This call](https://github.com/neovim/neovim/pull/5226/commits/5c8cbd52d0acb6c9ac28ca94c7b4fb80e5033a03#diff-4dbe255ebfc94f6664c66ffeab750809R542) seems not to be processed locally, but seems to be executed as 2 keypresses in some quickbuild runs. Maybe that's some indeterminism about faster/slower environments? What can I do about this? I tried inserting `wait()` inbetween the 2 keys, but then the `<bs>` would not be processed.

The test should actually fail, because these 2 keypresses, while leading to the same command line, will change the screen state (I tested manually).

KillTheMule commented:
Opened a vim issue: https://github.com/vim/vim/issues/983 about substitution an erroring in a specific test.

bfredl commented:
patch for buffer numbers:

```vim
diff --git a/src/nvim/ex_cmds.c b/src/nvim/ex_cmds.c
index e2c1768..8bc0eae 100644
--- a/src/nvim/ex_cmds.c
+++ b/src/nvim/ex_cmds.c
@@ -6152,7 +6152,7 @@ void do_inc_sub(exarg_T *eap)
   update_screen(0);
   if (livebuf != NULL && buf_valid(livebuf)) {
     close_windows(livebuf, false);
-    close_buffer(NULL, livebuf, DOBUF_WIPE, false);
+    wipe_buffer(livebuf, false);
   }
   unblock_autocmds();
   redraw_later(SOME_VALID);
```

note #4934 will bork this, but I will fix that once #4934 gets merged. Or if this is merged before, I'll include the fix in #4934.

KillTheMule commented:
Ok, nice, I integrated your patch and wrote a small test for it. I also replaced `klist` by `kvec`, it would be nice if someone could check [the commit](https://github.com/neovim/neovim/pull/5226/commits/e5e39da6ddd0fa6ba9e2858ad4c552465e6b7e46), especially the [deallocation part](https://github.com/neovim/neovim/pull/5226/commits/e5e39da6ddd0fa6ba9e2858ad4c552465e6b7e46#diff-b9d70f1e57978ae9fbf83f73315b0c0fR3933).

CI pending I'd say we're done once more.

bfredl commented on src/nvim/ex_cmds.h:
```diff
@@ -31,30 +32,19 @@ typedef struct {
-typedef klist_t(colnr_T) klist_colnr_T;
-#undef _noop
+// List of matches on a single line
+typedef kvec_t(colnr_T) kvec_colnr_T;
```
this typedef doesn't seem to buy much, just use `kvec_t(colnr_T)` in the code

ZyX-I replied:
It should be `ColNrVec` according to the style guide, not `…_T`.

@bfredl Using `kvec_t(colnr_T)` is not a good idea, you can’t pass `kvec_t(colnr_T) *foo` to `void bar(kvec_t(colnr_T) foo)`:

```
echo $'struct { int a; } *foo; int main(int argc, char **argv, char **environ){ struct { int a; } bar; foo = &bar; }' | clang -Wall -pedantic -std=c99 -xc -
<stdin>:1:101: warning: incompatible pointer types assigning to 'struct (anonymous struct at <stdin>:1:1) *' from 'struct (anonymous struct at <stdin>:1:74) *' [-Wincompatible-pointer-types]
struct { int a; } *foo; int main(int argc, char **argv, char **environ){ struct { int a; } bar; foo = &bar; }
                                                                                                    ^ ~~~~
1 warning generated.
```

bfredl replied:
@ZyX-I but we don't pass a pointer to the `kvec_t(colnr_T)`, it belongs to the enclosing MatchedLine struct, and we already use kvec_t without a typedef in this kind of case.

bfredl commented on src/nvim/ex_cmds.h:
```diff
@@ -31,30 +32,19 @@ typedef struct {
-KLIST_INIT(MatchedLine, MatchedLine, _dealloc_MatchedLine)
-typedef klist_t(MatchedLine) klist_MatchedLine;
-#undef _dealloc_MatchedLine
+typedef kvec_t(MatchedLine) kvec_MatchedLine;
```
ditto

KillTheMule replied:
The macro definition is [here](https://github.com/neovim/neovim/blob/master/src/nvim/lib/kvec.h#L52). Macros are just text replacements, so I'd redefine the struct everytime. Wouldn't that mean that after

```C
kvec_t(int) A;
kvec_t(int) B
```
the two variables A and B have a different type? Can I have this struct definitition in the signature of a function?

I'll give it a shot :)

KillTheMule replied:
Yeah no, it does not seem to work with the call to `ex_window_inc_sub`, unless I missed something obvious.

ZyX-I replied:
@KillTheMule Yes, they have different type, so `typedef` is needed. But neither `kvec_colnr_T` nor `kvec_MatchedLine` do not match style guide. I would rather suggest `ColNrs` (plural) or `ColNrVec` (`Vec` suffix) and same thing here unless you know a name that resembles this better.

KillTheMule replied:
No I'm really bad at naming (ever heard of Leonardo of Quirm?), so I'm happy to use your suggestions. I obviously just contracted `kvec_t(colnr_T)` to something without brackets...

bfredl commented on src/nvim/ex_cmds.h:
```diff
@@ -26,6 +29,26 @@ typedef struct {
+// Defs for inc_sub functionality
+
+// List of matches on a single line
+typedef kvec_t(colnr_T) ColNrVec;
```
this one you could still remove, as it's part of a bigger struct and not directly addressed, unlike `MatchedLineVec`

KillTheMule commented:
@bfredl So about that segfault, it's become quite strange. Firstly, I've added [a test](https://github.com/neovim/neovim/pull/5226/commits/f689de4bb443bfd70dbb043d501abcea164cb7a7#diff-4dbe255ebfc94f6664c66ffeab750809R738) that simply fails, that shows that the `n` flag does funky things with incsub. I suspect it is related.

Now, I could not make a test that segfaults. I _can_ however reproduce it reliably locally, by using these _exact_ steps:

- Open any file `:e tmp.txt`, it must contain the word "match" (doesn't matter what it is really, adjust the pattern accordingly). You can also directly load it from the command line.
- `:set ics=split` (nosplit, too)
- `:%s/match/X/gn` (does not need `gn`, only `n` suffices).

If you obtain the file text differently (copy&pase, type it, read it in via `:r tmp.txt`), the segfault does not occur.

bfredl commented:
isn't the problem simply that `n` does not do the substitution, but the flag that the substitution was done is still set, so we incorrectly attempt to undo-and-forget? (We also should have an error check in `undo_and_forget`, if there was nothing to undo, don't attempt to forget, instead of crashing)

KillTheMule commented:
I'm pretty sure that after I've been testing this in various circumstance for so long, such an easy explanation would have come even to the feeblest of minds... Well ****, you're right of course, and if one does an edit before, one can see that undone with the 'n' flag. I'm just thinking "Could be worse, the 'n' flag could be saved between substitutions like 'g'".

I can look at it tomorrow, if you don't beat me to it :)

jamessan commented:
> I'm just thinking "Could be worse, the 'n' flag could be saved between substitutions like 'g'".

Vim's 7.4.2219 patch could help with that, which I'm in the process of merging (c.f., #5231).

KillTheMule commented:
I was able to fix the problem with the `n` flag, that was easy \,,/

@bfredl: I'd need your help please for the error check in `u_undo_and_forget`, because I've not really understood the undo stuff. I tried [this](https://github.com/neovim/neovim/pull/5226/commits/d5317d001fe8d65296a4dabfd22dbe9d4b66cb86#diff-ac614bee43c84fee9c6ae32eeabfbcafR1690) which I got from [here](https://github.com/KillTheMule/neovim/blob/d5317d001fe8d65296a4dabfd22dbe9d4b66cb86/src/nvim/undo.c#L1756), but that basically made every test fail (generally speaking, typing the replacement text would show up in the command line but not in any of the 2 possible buffers).

bfredl commented:
Well, we will not be able to test it (in lua) as it should not ever happen (with incsub, if it happens, we need to change the code). I'll will look into adding the proper error check tomorrow. It is not really a blocker for incsub, it is more for when some other code will later try to use it, one will get a nice error message in this case instead of nvim segfaulting.

KillTheMule commented:
Great! So tomorrow will be squashing day, and after you've added the correct error check this is most definitely 100% positively ready (and while we're at it, I'd want a pony).

I'm marking it as RFC, since the remaining linter failures are either fixed locally already or will be so tomorrow.

bfredl commented:
Maybe like this

```diff
diff --git a/src/nvim/ex_cmds.c b/src/nvim/ex_cmds.c
index 2f45ff0..d651ef9 100644
--- a/src/nvim/ex_cmds.c
+++ b/src/nvim/ex_cmds.c
@@ -6149,7 +6149,9 @@ void do_inc_sub(exarg_T *eap)
   do_sub(eap);
   emsg_off--;
   if (sub_done == 1) {
-    u_undo_and_forget(1);
+    if (!u_undo_and_forget(1)) {
+      abort();
+    }
     sub_done = 0;
   }
   // Put back eap in first state
diff --git a/src/nvim/undo.c b/src/nvim/undo.c
index 4df439f..37369f4 100644
--- a/src/nvim/undo.c
+++ b/src/nvim/undo.c
@@ -1685,7 +1685,9 @@ void u_redo(int count)
 }
 
 /// undo and forget.
-void u_undo_and_forget(int count)
+///
+/// @return false if undo was not possible
+bool u_undo_and_forget(int count)
 {
   if (curbuf->b_u_synced == false) {
     u_sync(true);
@@ -1694,6 +1696,11 @@ void u_undo_and_forget(int count)
   undo_undoes = true;
   u_doit(count, true);
 
+  if (curbuf->b_u_curhead == NULL) {
+    // nothing was undone.
+    return false;
+  }
+
   // Delete the current redo header
   // set the redo header to the next alternative branch (if any)
   // otherwise we will be in the leaf state
@@ -1717,6 +1724,7 @@ void u_undo_and_forget(int count)
     curbuf->b_u_seq_last--;
   }
   u_freebranch(curbuf, to_forget, NULL);
+  return true;
 }
 
 /*
```

so that a future consumer sees that failure is possible and need to decide what to do.

KillTheMule commented:
No that does not work, unfortunately. The moment one type a second delimiter nvim aborts. Note that without your patch, everything seems to be working alright.

Also, I don't get the undo logic. Don't we need to undo _prior_ to the second delimiter? Because we're just using that as a search, but the real `do_sub` already treats it as deletion. After the second delimiter, we want to keep the new state of the buffer, just not have it recorded in the undo history until enter is pressed, in which case it isn't a live command anymore and should be executed (i.e. not undone) and recorded.

bfredl commented:
I tried both with manual testing and the test suite, and it seems to work. Can you a test case?

The delimiters shouldn't matter, we should undo-and-forget exactly when `do_sub()` changed something. And for now it does not change the buffer when `is_live && !has_second_delim` which seems reasonable. Maybe we could eliminate `sub_done` and just check whether the undo state is "dirty" (whether there are not yet synced changes), or set `sub_done` exactly at whatever `ml_` call that actually changes the buffer.

KillTheMule commented:
> I tried both with manual testing and the test suite, and it seems to work. Can you a test case?

Err, really, are you up-to-date? I have some trival local changes I'm gonna push now, but that just changes formatting/comments.

If I apply your patch, it's as easy as `build/bin/nvim -u NONE`, `itw`, `:set ics=split`, and then `:%s/tw/`, and upon typing the slash stuff crashes (there's no core file, which I assume is because we abort?).

bfredl commented:
Yes, very up to date. Though in the latest push it looks like you added the check manually, but in the wrong place (at the start of the function, not after `u_doit` )

KillTheMule commented:
Oh Christ, I'll check.

KillTheMule commented:
@khalidchawtany:

>     @KillTheMule when doing a substitute and change font size in the terminal or resize the terminal window. The incsub window disappears till you make a change to the substitution string.
>     Is the size of incsub window fixed? It is always composed of 7 lines regardless of neovim window size.

The size is determind by `cmdwinheigt`, which defaults to 7. I can confirm it vanishing upon terminal resize, but I'd consider this a somewhat minor bug and would not block this PR because of it. I'm not sure how easily that would be fixed as well... @bfredl, thoughts?

KillTheMule commented:
Travis fails because of gcov. Needs a bigger rebase though, I'll get it done in the next few days. After that, I don't see why ~~this should not be merged~~ the next problem could not surface.

khalidchawtany commented:
@KillTheMule I agree that the vanishing is a minor bug. However, I cannot figure out how to change `cmdwinheight` while doing a live substitute.

bfredl commented:
> The size is determind by cmdwinheigt

Ideally this should controlled by a separate option, or even be controllable at cmdline-mode-time, so that the current preview window is resized. (While the "current" preview window of course does not exist, we could emulate that by making the override local to the current cmdline-mode activation and retrigger the DOCMD_LIVE_PREVIEW). But neither would I consider it a blocker for this PR.

KillTheMule commented:
> @KillTheMule I agree that the vanishing is a minor bug. However, I cannot figure out how to change cmdwinheight while doing a live substitute.

Yeah, you can't. Your usecase would be to adjust to the number of matches? I'd hate that, but as you can see, bfredl already has ideas :)

KillTheMule committed:
Linted

KillTheMule commented:
Rebased on current master.

timeyyy commented:
Ref #4794

balta2ar commented:
@KillTheMule May I inquire what's the status here and what else needs to be done in order to merge it? I'm really looking forward to get my feed wet with this feature.

KillTheMule commented:
From my POV this is ready, needs a review.

bfredl commented on src/nvim/ex_cmds.c:
```diff
@@ -64,6 +65,9 @@
 typedef struct sign sign_T;
 
+// boolean to know if we have to undo
+static int sub_done = 0;
```
bool, also needs a better name, maybe `inc_sub_needs_undo` or `incsub_did_changes`

bfredl commented on src/nvim/ex_cmds.c:
```diff
@@ -3112,6 +3118,9 @@ void do_sub(exarg_T *eap)
   int start_nsubs;
   int save_ma = 0;
 
+  sub_done = 0;
```
IIUC `sub_done` should always be false here anyway

KillTheMule replied:
No, it can be `true` here when changing the `ics` setting from `""` to something else. We could reset it somewhere else (say, [here](https://github.com/neovim/neovim/pull/5226/files#diff-b9d70f1e57978ae9fbf83f73315b0c0fR6188)), or make it a "nontritivial `ics` setting only" variable, but the code would get more complicated. I'd vote for keeping it the way it is.

bfredl commented on src/nvim/ex_cmds.c:
```diff
@@ -3237,7 +3251,9 @@ void do_sub(exarg_T *eap)
   }
 
-  if (search_regcomp(pat, RE_SUBST, which_pat, SEARCH_HIS, &regmatch) == FAIL) {
+  int searchRC_option = eap->is_live ? 0 : SEARCH_HIS;
```
maybe just `search_options`

bfredl commented on src/nvim/ex_cmds.c:
```diff
@@ -5935,3 +6036,191 @@ void set_context_in_sign_cmd(expand_T *xp, char_u *arg)
   }
 }
+
+/// Open a window for future displaying of the inc_sub mode.
```
remove "future", is displayed right now :)

bfredl commented on src/nvim/ex_cmds.c:
```diff
@@ -5935,3 +6036,191 @@ void set_context_in_sign_cmd(expand_T *xp, char_u *arg)
+/// Open a window for future displaying of the inc_sub mode.
+///
+/// Does not allow editing in the window.
+/// Returns when the window is closed.
```
This sounds almost like "wait until the user [or some other code] closes the window, then return". Maybe "closes the window and restores the window layout before returning"

bfredl commented on src/nvim/ex_cmds.c:
```diff
@@ -5935,3 +6036,191 @@ void set_context_in_sign_cmd(expand_T *xp, char_u *arg)
+
+  redraw_later(SOME_VALID);
+
+  // Restore the old window
```
maybe this also should be in `if(split)` just to make it clear this restores the state from window splitting.

bfredl commented on src/nvim/ex_cmds.c:
```diff
@@ -5935,3 +6036,191 @@ void set_context_in_sign_cmd(expand_T *xp, char_u *arg)
+/// @param pat    The pattern word
+/// @param sub    The replacement word
+/// @param lmatch The list containing our data
+void ex_window_inc_sub(char_u * pat,
```
I believe this function is responsible for trigger the right redraws even when a window is not opened `incsub=nosplit`. Maybe `static void inc_sub_display`.

bfredl commented on src/nvim/ex_cmds.c:
```diff
@@ -5935,3 +6036,191 @@ void set_context_in_sign_cmd(expand_T *xp, char_u *arg)
+
+/// :substitute command implementation
+///
+/// Uses do_sub() to do the actual substitution. Calls :u after each typed
```
It doesn't call `:u` anymore really. Maybe "undoes the substitution (and removes it from undo history) if the text changed, unless finishing the command"

bfredl commented on src/nvim/ex_cmds.c:
```diff
@@ -5935,3 +6036,191 @@ void set_context_in_sign_cmd(expand_T *xp, char_u *arg)
+    if (!u_undo_and_forget(1)) {
+      abort();
+    }
+  sub_done = 0;
```
indentation

bfredl commented:
After rebasing on master [this patch](https://github.com/bfredl/neovim/commit/6b03d01b5c014bc1c5bdcc54489c4c41fa60b902) could be used to reuse a single bufnr. It's admittedly a bit ugly. An option could be to bdelete the buffer instead of bwiping it, but then the user could revive the buffer themselves and do all sort of things to it, causing all sorts of trouble on the next incsub... 

Apart of that, and some minor stylistic issues (I will take a look at the tests later also), I think this is more or less ready. As I think I said before, there will likely be more bugs, but exposing  it to more users might be necessary to find them all. As long as there is no bugs when incsub is inactive, merging this should be ok from a stability point of view.

KillTheMule commented:
Thanks, bfredl. Small headsup: I'll probably only be able to work on it this weekend, sorry for the delay!

KillTheMule commented:
I did all the edits an rebased on master, applied your patch... I can't compile locally right now (asking about it on gitter), so I just pushed and hope for the best :)

KillTheMule commented:
Oh, ok, then I was a bit quick to assume there's I problem. I just saw "bufnr is 3 instead of 2" and assumed that's a mistake. I'll see if I can make the test work again.

KillTheMule commented:
Ok, the test passes after adjusting the buffer number. I've also rebase on master, so let's pray CI passes, after which I think this could be merged.

KillTheMule commented:
I had to adjust [`buffer_spec.lua`](https://github.com/neovim/neovim/pull/5226/files#diff-dc6511dd3a11b287883b371b18533fa5R16), the tests do pass now, but please have a look anyways @bfredl. Does this added argument need additional testing maybe? We could put up an entry-level Issue I'd say...

KillTheMule commented:
Travis failures unrelated.

Appveyor failures are genuine incsub test failures. The original window seems to scroll differently here, scrolls down 1 more line than on linux etc... I have no idea how to diagnose this. @equalsraf is that something known on windows?

equalsraf commented:
@KillTheMule the reason why no screen tests are enabled in Appveyor is because there are subtle differences when drawing, e.g. text gets elided earlier than in Linux. Unfortunately I do not know the cause.

KillTheMule commented:
Thanks, so should I disable those screen tests in windows? Is there a generic way to do that?

equalsraf commented:
> Thanks, so should I disable those screen tests in windows? Is there a generic way to do that? 

There is ([example](https://github.com/neovim/neovim/blob/0ade1bb7067d0cdb9b059fb66a8c4b868408be9c/test/functional/options/shortmess_spec.lua#L5)). Add the following to the spec file or to the test

```lua
if helpers.pending_win32(pending) then return end
```

KillTheMule commented:
Ah thanks, I'll do that.

bfredl commented on test/functional/ui/incsubstitution_spec.lua:
```diff
@@ -0,0 +1,1214 @@
+    execute("set nohlsearch")
+    execute("hi IncSubstitute guifg=red guibg=yellow")
+    screen:attach()
+    screen:set_default_attr_ignore( {{bold=true, foreground=Screen.colors.Blue}} )
```
please remove this, by the new standard #5197  we prefer to not ignore anything, unless there's a really good reason (and thanks to `snapshot_util()` there typically isn't)

bfredl commented on test/functional/ui/incsubstitution_spec.lua:
```diff
@@ -0,0 +1,1214 @@
+    })
+  end
+
+  if incsub then
```
is this `if` needed? can't we make sure that incsub always is the empty string if it should be off.

KillTheMule replied:
No, sure, we control all the calls. I usually find it a good practice to have nil work as an argument as well without failure, and in this case the intention would be pretty clear. Not married to that thought, though, so I'll remove it if you wish.

bfredl commented on test/functional/ui/incsubstitution_spec.lua:
```diff
@@ -0,0 +1,1214 @@
+    screen:detach()
+  end)
+
+  it('with undolevels=-1', function()
```
Maybe also test undolevels=-1 with the substitution being aborted.

bfredl reviewed (commented):
Looks good, just some minor suggestions.

KillTheMule commented:
Ok, rebasing on master broke all the stuff :( I fixed the tests as far as that was useful:

- The [behavior with a failed mapping](https://github.com/neovim/neovim/pull/5226/files#diff-4dbe255ebfc94f6664c66ffeab750809R870) is now homogeneous, which is nothing to fret about I'd say
- The scrolling behavior of the original window changed, which now is more in-line with the original substitution, which is a good thing imho

The real problem is that the incsub window is not shown anymore in the tests. Seems to be an issue of the test setup though, because when trying manually everything works as expected. @bfredl @justinmk any idea what's causing this? As an example, see the first test failure on CI. Also, highlighting does not seem to work properly, but that might be related.

@bfredl I will try to add the new test you requested. I'll change ignoring the highlights after we've sorted out the other test failures.

KillTheMule commented:
Skip my last comment, I'm simply very stupid. All works now, I've also removed the ignored highlight, and fixed a bug found by clang that really makes me think "How could this ever have worked?". Hope all tests pass now :)

KillTheMule commented:
I added the pending calls for windows. The problem I faced is that you can't call `pending` in an `it` block, you need to do it in a `describe`, which basically means almost no test runs on windows, even though there are several that do not check the screen state.

Other than that, this is ready once more, hopefully CI passes \**crosses fingers*\*

KillTheMule commented:
Hoorayyy... ping @bfredl, think this is mergeable? I did only touch the tests since last time... I should probably merge those 2 commits, the original intention is pretty much watered down by now. Anyways, I hope we can lift this into the wild now :)

bfredl commented:
To me this is very mergable. But I want to recall @justinmk saying something like "this is one of the first things we should get merged after 0.1.6", but I could remember wrong. Anyway, I think the plan is to release 0.1.6 very soon, after the API versioning (#5535) is done or so.

jamessan commented on runtime/doc/syntax.txt:
```diff
@@ -4855,6 +4855,8 @@ IncSearch	'incsearch' highlighting; also used for the text replaced with
 							*hl-LineNr*
 LineNr		Line number for ":number" and ":#" commands, and when 'number'
 		or 'relativenumber' option is set.
+                                                        *hl-IncSubstitute*
```
Why is this indented differently than the rest of the `hl-*` topics?

KillTheMule replied:
There's a secret and really smart reason, but if I told you, I'd have to kill you.

So I'm just gonna fix that :D

jamessan commented on test/functional/ui/incsubstitution_spec.lua:
```diff
@@ -0,0 +1,1276 @@
+      return
+    end
+
+--    for _, case in pairs(cases) do
```
Since this test is always marked pending, can the test code be uncommented?

KillTheMule replied:
Sure thing!

KillTheMule replied:
See below, the linter will be unhappy.

KillTheMule commented:
> But I want to recall @justinmk saying something like "this is one of the first things we should get merged after 0.1.6", but I could remember wrong

Sure, there's no expectation from my side that this gets merged fast or something (I certainly had to take my time about this lately). I just want to get this into the "ready to merge" state, and it's certainly not ready unless _you_ say so :)

KillTheMule commented:
Ok, did the last few changes, should be good now, travis failure seems unrelated (32bit build hung). I'm marking this RDY for now.

Jamessan, I could not uncomment that text, the linter screams "Unreachable code" at me, and I'd like to keep the linter clean. If vim fixes that behavior and we port it, manually enabling the test is needed anyways, so it's not a huge loss :)

justinmk commented:
Noticed a small bug, not going to block this though: after adding `/g` to match all patterns in a line, the IncSearch highlight position does not get updated. Steps:

Insert some text like this:

    foo bar foo
    foo bar foo
    foo bar foo

with `incsub=split`, do this:

    :%s/\vfoo/XXXXXX/

All good so far. Now add `/g`:

    :%s/\vfoo/XXXXXX/g

The `IncSearch` highlight position is wonky.

KillTheMule commented:
I can confirm. Note that that without the `\v`, it works nicely. With the `\v`, the highlighting of the second appearance is 2 spaces to the left, so maybe there's just a small bug computing the width of the search pattern.

justinmk commented on runtime/doc/options.txt:
```diff
@@ -3991,6 +3992,16 @@ A jump table for the options with a short description can be found at |Q_op|.
+	while the user types the command.  If set to "split", a split window
+	is open which displays the lines where the search matches. The
+	replacement text in the split is hightlighted using
+	|hl-IncSubstitute|. Note: Only '/' is supported as a delimiter.
```
AFAICT all delimiters are supported now. 
Also this help entry should be located next to 'incsearch'.

I'll update this when merging.

justinmk commented on runtime/syntax/vim.vim:
```diff
@@ -58,7 +58,7 @@ syn case ignore
 
 " Default highlighting groups {{{2
-syn keyword vimHLGroup contained	ColorColumn Cursor CursorColumn CursorIM CursorLine CursorLineNr DiffAdd DiffChange DiffDelete DiffText Directory ErrorMsg FoldColumn Folded IncSearch LineNr MatchParen Menu ModeMsg MoreMsg NonText Normal Pmenu PmenuSbar PmenuSel PmenuThumb Question Scrollbar Search SignColumn SpecialKey SpellBad SpellCap SpellLocal SpellRare StatusLine StatusLineNC TabLine TabLin ...
+syn keyword vimHLGroup contained	ColorColumn Cursor CursorColumn CursorIM CursorLine CursorLineNr DiffAdd DiffChange DiffDelete DiffText Directory ErrorMsg FoldColumn Folded IncSearch IncSubstitute LineNr MatchParen Menu ModeMsg MoreMsg NonText Normal Pmenu PmenuSbar PmenuSel PmenuThumb Question Scrollbar Search SignColumn SpecialKey SpellBad SpellCap SpellLocal SpellRare StatusLine StatusLineNC  ...
```
should be in `nvimHLGroup`

KillTheMule commented:
Uhh after a quick check, things look more dire than I thought. The whole incsub window does not account for multiple matches on a line, let alone multiline matches. I assume fixing that might take some thought...

justinmk commented on src/nvim/buffer.c:
```diff
@@ -1335,11 +1336,12 @@ void do_autochdir(void)
+    char_u *sfname,            // short fname or NULL
+    linenr_T lnum,             // preferred cursor line
+    int flags,                 // BLN_ defines
+    handle_T bufnr
```
instead of this, was there discussion about maybe using `BLN_DUMMY`? (still investigating this myself)

bfredl replied:
I didn't see `BLN_DUMMY` avoiding allocating a buffer number?

bfredl replied:
> Can we instead use the same ffname every time, so that buflist_new() finds the same existing incsub buffer?

No we wipe the buffer when we're done. If the buffer exists then the user could mess up the buffer by changing settings etc until the next time.

> Hmm, even q: and :copen use a new bufnr each time.

but "each time" is a lot more often for incsub, almost every keypress after `:%s`.

justinmk replied:
Hacking around, it looks like the code can be simplified a good amount by "not worrying" about the user accidentally editing this special buffer. I'd like to try to solve that more general problem rather than add the special cases for incsub buffer. There is already some internal notion in Vim for "special buffers", which we maybe can extend and formalize a bit instead of adding new parameters.

Tests are passing, I'll push a PR on top of this PR later today.

bfredl commented on src/nvim/buffer.c:
```diff
@@ -1458,7 +1460,9 @@ buflist_new (
-    buf->b_fnum = top_file_num++;
+    // If bufnr is nonzero it is assumed to be a previously
+    // reserved buffer number (handle)
+    buf->handle = bufnr != 0 ? bufnr : top_file_num++;
```
They are synonymous, but probably better `b_fnum` yes (I reckon we won't change `b_fnum` to `handle` generally, to avoid friction merging vim patches)

justinmk commented on src/nvim/ex_docmd.c:
```diff
@@ -271,7 +271,7 @@ do_exmode (
 {
   return do_cmdline((char_u *)cmd, NULL, NULL,
-      DOCMD_VERBOSE|DOCMD_NOWAIT|DOCMD_KEYTYPED);
+                    DOCMD_NOWAIT|DOCMD_KEYTYPED);
```
`DOCMD_VERBOSE` was removed in several places. This wasn't in the original PR, was it discussed somewhere, or could you explain?

zeertzjq replied:
Is there a conclusion about why this was removed from `do_cmdline_cmd()`?

zeertzjq replied:
I've created a PR #23482 to revert this change.
